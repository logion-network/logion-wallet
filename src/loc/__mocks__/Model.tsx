import {
    rejectProtectionRequest,
    acceptProtectionRequest,
} from './ModelMock';

export {
    rejectProtectionRequest,
    acceptProtectionRequest,
};
