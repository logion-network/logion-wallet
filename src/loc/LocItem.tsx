import {
    LogionClient,
    LocRequestState,
    LocData,
    LocsState,
    EditableRequest,
    MergedLink,
    PublicLoc,
    ItemStatus,
    ReviewResult,
    HashString,
    OpenLoc,
    PendingRequest,
} from "@logion/client";
import { UUID, LocType, Fees, ValidAccountId, Hash } from "@logion/node-api";
import { ReactNode, useCallback } from "react";
import { CallCallback } from "src/ClientExtrinsicSubmitter";
import { Viewer } from "src/common/CommonContext";
import { isGrantedAccess } from "src/common/Model";
import { Rules } from "src/common/Responsive";
import SubmitterName from "src/common/SubmitterName";
import { Cell, Column, DateTimeCell } from "src/common/Table";
import { Child } from "src/common/types/Helpers";
import ViewFileButton from "src/common/ViewFileButton";
import { documentClaimHistoryPath } from "src/legal-officer/LegalOfficerPaths";
import { fullCertificateUrl } from "src/PublicPaths";
import { documentClaimHistoryPath as userDocumentClaimHistoryPath } from "src/wallet-user/UserRouter";
import LocLinkDetails from "./LocLinkDetails";
import LocPrivateFileDetails from "./LocPrivateFileDetails";
import LocPublicDataDetails from "./LocPublicDataDetails";
import RestrictedDeliveryCell from "./RestricedDeliveryCell";
import { ContributionMode } from "./types";
import ReviewStatusCell from "./ReviewStatusCell";
import HelpTooltip from "src/components/helptooltip/HelpTooltip";
import { DateTime } from "luxon";

export type LocItemType = 'Data' | 'Document' | 'Linked LOC';

export interface LinkData {
    readonly linkedLoc: LocData;
    readonly linkDetailsPath: string;
    readonly nature: HashString;
}

export interface MetadataData {
    readonly name: HashString;
    readonly value: HashString;
}

export interface FileData {
    readonly hash: Hash;
    readonly fileName: string;
    readonly nature: HashString;
    readonly size: bigint;
    readonly storageFeePaidBy: string;
}

export interface CommonData {
    readonly timestamp: DateTime | null;
    readonly type: LocItemType;
    readonly submitter?: ValidAccountId;
    readonly status: ItemStatus;
    readonly newItem: boolean;
    readonly template: boolean;
    readonly isSet?: boolean;
    readonly fees?: Fees;
    readonly reviewedOn?: DateTime;
    readonly rejectReason?: string;
    readonly acknowledgedByOwner?: boolean;
    readonly acknowledgedByVerifiedIssuer?: boolean;
    readonly defaultTitle?: string;
}

abstract class AbstractLocItem<T> implements CommonData {

    constructor(commonData: CommonData, data?: T) {
        this.commonData = commonData;
        this._data = data;
    }

    protected commonData: CommonData;
    private readonly _data?: T;

    get timestamp() {
        return this.commonData.timestamp;
    }

    get type() {
        return this.commonData.type;
    }

    get submitter() {
        return this.commonData.submitter;
    }

    get status() {
        return this.commonData.status;
    }

    get newItem() {
        return this.commonData.newItem;
    }

    get template() {
        return this.commonData.template;
    }

    get isSet() {
        return this.commonData.isSet;
    }

    get fees() {
        return this.commonData.fees;
    }

    get reviewedOn() {
        return this.commonData.reviewedOn;
    }

    get rejectReason() {
        return this.commonData.rejectReason;
    }

    get acknowledgedByOwner() {
        return this.commonData.acknowledgedByOwner;
    }

    get acknowledgedByVerifiedIssuer() {
        return this.commonData.acknowledgedByVerifiedIssuer;
    }

    get defaultTitle(): string  {
        return this.commonData.defaultTitle || this.commonData.type;
    }

    isPublishedOrAcknowledged(): boolean {
        return this.commonData.status === "PUBLISHED" || this.commonData.status === "ACKNOWLEDGED"
    }

    abstract title(): string;

    hasData(): boolean {
        return this.commonData.submitter !== undefined && this._data !== undefined;
    }

    data(): T {
        if(this._data) {
            return this._data;
        } else {
            throw new Error("Data not set");
        }
    }

    as<U>(): U {
        return this._data as U;
    }

    abstract publish(timestamp: DateTime | null, fees?: Fees, storageFeePaidBy?: string): AbstractLocItem<T>;

    get submitterOrThrow() {
        if(this.commonData.submitter) {
            return this.commonData.submitter;
        } else {
            throw new Error("Data not set");
        }
    }
}

export class MetadataItem extends AbstractLocItem<MetadataData> {

    static TYPE: LocItemType = "Data";

    override title() {
        if (this.hasData()) {
            return this.data().name.value || "-";
        } else {
            return this.defaultTitle;
        }
    }

    override publish(timestamp: DateTime | null, fees?: Fees, _storageFeePaidBy?: string): MetadataItem {
        return new MetadataItem(
            {
                ...this.commonData,
                timestamp,
                fees,
            },
            this.data(),
        );
    }
}

export class FileItem extends AbstractLocItem<FileData> {

    override title() {
        if (this.hasData()) {
            return this.data().nature.value || "-";
        } else {
            return this.defaultTitle
        }
    }

    override publish(timestamp: DateTime | null, fees?: Fees, storageFeePaidBy?: string): FileItem {
        const newFileData = this.data() && storageFeePaidBy ? {
            ...this.data(),
            storageFeePaidBy,
        } : this.data();
        return new FileItem(
            {
                ...this.commonData,
                timestamp,
                fees,
            },
            newFileData,
        );
    }
}

export class LinkItem extends AbstractLocItem<LinkData> {

    override title() {
        if (this.hasData()) {
            return this.data().nature.value || "-";
        } else {
            return this.defaultTitle;
        }
    }

    override publish(timestamp: DateTime | null, fees?: Fees, _storageFeePaidBy?: string): LinkItem {
        return new LinkItem(
            {
                ...this.commonData,
                timestamp,
                fees,
            },
            this.data(),
        );
    }
}

export type LocItem = LinkItem | MetadataItem | FileItem;

export enum PublishStatus {
    NONE,
    START,
    PUBLISHING,
    DONE
}

export interface PublishState {
    status: PublishStatus;
}

export interface PublishProps {
    locItem: LocItem;
    itemType: 'Public Data' | 'Document' | 'Link';
    publishMutator: (current: LocRequestState, callback: CallCallback) => Promise<LocRequestState>;
    feesEstimator: () => Promise<Fees>;
}

export function buildItemTableColumns(args: {
    loc: LocData,
    locState: LocRequestState,
    width: (rule: Rules) => string,
    currentAddress: string | undefined,
    contributionMode: ContributionMode | undefined,
    viewer: Viewer,
    renderActions: (locItem: LocItem) => ReactNode,
}): Column<LocItem>[] {
    const {
        loc,
        locState,
        width,
        currentAddress,
        contributionMode,
        viewer,
        renderActions,
    } = args;
    let columns: Column<LocItem>[] = [
        {
            header: "Public description",
            render: locItem => <Cell content={ locItem.title() } overflowing tooltipId={`${loc.id}-name-tooltip`}/>,
            renderDetails: locItem => renderDetails(loc, locItem, viewer),
            detailsExpanded: locItem => locItem.newItem || (locItem.template && (viewer !== "LegalOfficer" || (locItem.status !== "PUBLISHED" && locItem.status !== "ACKNOWLEDGED"))),
            hideExpand: locItem => locItem.template && viewer !== "LegalOfficer",
            align: "left",
            width: width({
                onSmallScreen: '150px',
                otherwise: '250px'
            }),
        },
        {
            header: "Timestamp",
            render: locItem => <DateTimeCell
                dateTime={ locItem.timestamp }
                spinner={ locItem.status === 'PUBLISHED' } />,
            width: width({
                onSmallScreen: '100px',
                otherwise: '200px'
            }),
        },
        {
            header: "Type",
            render: locItem => <Cell content={
                <>
                    <span className="item-type">{ locItem.type }</span> {
                        locItem.type === 'Document' &&
                        canViewFile(viewer, currentAddress, locItem, contributionMode) &&
                        <ViewFileButton
                            nodeOwner={ loc.ownerAddress }
                            fileName={ locItem.as<FileData>().fileName }
                            downloader={ () => locState?.getFile(locItem.as<FileData>().hash) }
                        />
                    }
                </> } />,
            width: "160px",
            align: "left"
        },
        {
            header: "Submitted by",
            render: locItem => <SubmitterName loc={ loc } submitter={ locItem.submitter?.address } />,
        }
    ];

    if(loc.locType === "Collection" && viewer === "LegalOfficer") {
        columns.push({
            header: "Restricted Delivery?",
            render: locItem => locItem.type === "Document" && locItem.hasData() ? <RestrictedDeliveryCell hash={ locItem.as<FileData>().hash }/> : null,
            width: "130px",
        });
    }

    columns = columns.concat([
        {
            header: <>Reviewed <HelpTooltip id="reviewed" help="You confirm that the content you submitted for review by the Legal Officer is good to be recorded by the logion infrastructure with regard to the LOC context and to the specific content that will be publicly available after the recording (eg: GDPR)."/></>,
            render: locItem => <ReviewStatusCell locItem={ locItem } />,
        },
        {
            header: "",
            render: locItem => renderActions(locItem),
            width: width({
                onSmallScreen: viewer === "LegalOfficer" ? '345px' : '145px',
                otherwise: '400px'
            }),
        }
    ]);
    return columns;
}

function renderDetails(loc: LocData | undefined, locItem: LocItem, viewer: Viewer): Child {
    return (
        <>
            { locItem.type === 'Data' && <LocPublicDataDetails item={ locItem as MetadataItem } /> }
            {
                locItem.type === 'Document' &&
                <LocPrivateFileDetails
                    item={ locItem as FileItem }
                    documentClaimHistory={ loc?.locType === "Collection" && !locItem.template ? documentClaimHistory(viewer, loc, locItem.as<FileData>().hash) : undefined }
                    otherFeesPaidByRequester={ loc?.requesterLocId === undefined }
                />
            }
            { locItem.type === 'Linked LOC' && <LocLinkDetails item={ locItem as LinkItem } /> }
        </>
    )
}

function canViewFile(viewer: Viewer, address: string | undefined, item: LocItem, contributionMode?: ContributionMode): boolean {
    return item.hasData() && (contributionMode !== 'VerifiedIssuer' || item.submitterOrThrow.address === address) && (viewer === "User" || item.status !== "DRAFT");
}

export function documentClaimHistory(viewer: Viewer, loc: LocData | null, hash: Hash) {
    if(!loc) {
        return "";
    } else if(viewer === "LegalOfficer") {
        return documentClaimHistoryPath(loc.id, hash);
    } else {
        return userDocumentClaimHistoryPath(loc.id, hash);
    }
}

export function useDeleteMetadataCallback(mutateLocState: (mutator: (current: LocRequestState) => Promise<LocRequestState | LocsState>) => Promise<void>) {
    return useCallback(async (item: LocItem) => {
        await mutateLocState(async current => {
            if(current instanceof EditableRequest) {
                return current.deleteMetadata({
                    nameHash: item.as<MetadataData>().name.hash,
                });
            } else {
                return current;
            }
        });
    }, [ mutateLocState ]);
}

export function useDeleteFileCallback(mutateLocState: (mutator: (current: LocRequestState) => Promise<LocRequestState | LocsState>) => Promise<void>) {
    return useCallback(async (item: LocItem) => {
        await mutateLocState(async current => {
            if(current instanceof EditableRequest) {
                return current.deleteFile({
                    hash: item.as<FileData>().hash,
                });
            } else {
                return current;
            }
        });
    }, [ mutateLocState ]);
}

export function useDeleteLinkCallback(mutateLocState: (mutator: (current: LocRequestState) => Promise<LocRequestState | LocsState>) => Promise<void>) {
    return useCallback(async (item: LocItem) => {
        await mutateLocState(async current => {
            if(current instanceof EditableRequest) {
                return current.deleteLink({
                    target: item.as<LinkData>().linkedLoc.id,
                });
            } else {
                return current;
            }
        });
    }, [ mutateLocState ]);
}

export function canDelete(account: ValidAccountId | undefined, item: LocItem, viewer: Viewer, loc: LocData): boolean {
    return item.submitter?.address === account?.address && item.submitter?.type === account?.type
            && (loc.status === "DRAFT" || loc.status === "OPEN")
            && (item.status === "DRAFT" || item.status === "REVIEW_REJECTED");
}

export function canAdd(viewer: Viewer, loc: LocData) {
    return (viewer === "User" && (!loc.voidInfo && (loc.status === "DRAFT" || loc.status === "OPEN")))
        || (viewer === "LegalOfficer"
            && (!loc.voidInfo && loc.status === "OPEN")
            && ( loc.requesterAddress === undefined || loc.requesterAddress.type !== "Polkadot"));
}

type ActorType = "Owner" | ContributionMode;

function getActorType(viewer: Viewer, contributionMode?: ContributionMode): ActorType {
    if (viewer === "LegalOfficer") {
        return "Owner"
    } else {
        return contributionMode!
    }
}

function getSubmitterType(loc: LocData, item: LocItem): ActorType {
    const submittedByOwner = item?.submitter?.address === loc.ownerAddress && item?.submitter?.type === "Polkadot";
    const submittedByRequester = loc.requesterAddress !== undefined && item.submitter !== undefined && loc.requesterAddress.equals(item.submitter);

    if (submittedByOwner) {
        return "Owner";
    } else if (submittedByRequester) {
        return "Requester";
    } else {
        return "VerifiedIssuer";
    }
}

export function canPublish(viewer: Viewer, loc: LocData, item: LocItem, contributionMode?: ContributionMode): boolean {

    const actorType = getActorType(viewer, contributionMode);
    if (actorType === "VerifiedIssuer") {
        return false;
    }

    const submitterType = getSubmitterType(loc, item);

    const publishedByOwner = submitterType === "Owner" || loc.requesterAddress === undefined || loc.requesterAddress.type !== "Polkadot";

    const publishable = item.status === "REVIEW_ACCEPTED";

    return publishable
        && loc.status === "OPEN" && !loc.voidInfo
        && (
            (actorType === "Owner" && publishedByOwner) ||
            (actorType === "Requester" && !publishedByOwner)
        );
}

export function canAcknowledge(viewer: Viewer, loc: LocData, item: LocItem, contributionMode?: ContributionMode): boolean {

    const actorType = getActorType(viewer, contributionMode);
    if (actorType === "Requester") {
        return false;
    }

    const submitterType = getSubmitterType(loc, item);

    const acknowledgeable = item.status === "PUBLISHED";

    return acknowledgeable
        && loc.status === "OPEN" && !loc.voidInfo
        && (
            (submitterType === "VerifiedIssuer" && actorType === "VerifiedIssuer" && item.acknowledgedByVerifiedIssuer !== undefined && !item.acknowledgedByVerifiedIssuer)
            || (actorType === "Owner" && item.acknowledgedByOwner !== undefined && !item.acknowledgedByOwner)
        )
}

export function canRequestReview(viewer: Viewer, loc: LocData, item: LocItem) {
    return viewer === "User" && item.status === "DRAFT" && loc.status === "OPEN";
}

export function canReview(viewer: Viewer, loc: LocData, item: LocItem) {
    return viewer === "LegalOfficer" && item.status === "REVIEW_PENDING" && (loc.status === "REVIEW_PENDING" || loc.status === "OPEN");
}

export async function getLinkData(
    address: string | undefined,
    locsState: LocsState,
    link: MergedLink,
    detailsPath: (id: UUID, locType: LocType) => string,
    client: LogionClient,
): Promise<LinkData> {
    let linkedLoc: LocData | undefined;
    const targetId = link.target;
    let linkedLocState: LocRequestState | PublicLoc | undefined = locsState.findByIdOrUndefined(targetId);
    if (!linkedLocState) {
        linkedLocState = await client.public.findLocById({ locId: targetId });
        if(linkedLocState) {
            linkedLoc = linkedLocState.data;
        }
    } else {
        linkedLoc = linkedLocState.data();
    }

    if(!linkedLoc) {
        throw new Error("Unable to locate linked LOC");
    }

    let linkDetailsPath: string;
    if (isGrantedAccess(address, linkedLoc)) {
        linkDetailsPath = detailsPath(linkedLoc.id, linkedLoc.locType);
    } else {
        linkDetailsPath = fullCertificateUrl(linkedLoc.id);
    }
    return {
        linkedLoc,
        linkDetailsPath,
        nature: link.nature,
    };
}

export function useRequestReviewCallback(mutateLocState: (mutator: (current: LocRequestState) => Promise<LocRequestState | LocsState>) => Promise<void>) {
    return useCallback(async (locItem: LocItem) => {
        mutateLocState(async current => {
            if (current instanceof EditableRequest) {
                if (locItem.type === "Data") {
                    return current.requestMetadataReview({ nameHash: locItem.as<MetadataData>().name.hash });
                } else if (locItem.type === "Document") {
                    return current.requestFileReview({ hash: locItem.as<FileData>().hash });
                } else if (locItem.type === "Linked LOC") {
                    return current.requestLinkReview({ target: locItem.as<LinkData>().linkedLoc.id });
                } else {
                    return current;
                }
            } else {
                return current;
            }
        });
    }, [ mutateLocState ]);
}

export function useReviewCallback(mutateLocState: (mutator: (current: LocRequestState) => Promise<LocRequestState | LocsState>) => Promise<void>) {
    return useCallback(async (locItem: LocItem, decision: ReviewResult, rejectReason?: string) => {
        mutateLocState(async current => {
            if(current instanceof OpenLoc || current instanceof PendingRequest) {
                if(locItem.type === "Data") {
                    return current.legalOfficer.reviewMetadata({
                        nameHash: locItem.as<MetadataData>().name.hash,
                        decision,
                        rejectReason
                    });
                } else if(locItem.type === "Document") {
                    return current.legalOfficer.reviewFile({
                        hash: locItem.as<FileData>().hash,
                        decision,
                        rejectReason
                    });
                } else if(locItem.type === "Linked LOC") {
                    return current.legalOfficer.reviewLink({
                        target: locItem.as<LinkData>().linkedLoc.id,
                        decision,
                        rejectReason
                    });
                } else {
                    return current;
                }
            } else {
                return current;
            }
        });
    }, [ mutateLocState ]);
}
