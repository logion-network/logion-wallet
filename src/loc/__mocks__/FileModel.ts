export { getAllDeliveries, getAllCollectionDeliveries } from './FileModelMock';
