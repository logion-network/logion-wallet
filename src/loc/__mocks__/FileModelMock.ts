export let getAllDeliveries = jest.fn().mockResolvedValue({});
