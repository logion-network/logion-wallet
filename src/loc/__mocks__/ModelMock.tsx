export let rejectRequest = jest.fn();

export let acceptRequest = jest.fn();

export function setAcceptRequest(mockFn: any) {
    acceptRequest = mockFn;
}

export let setAssetDescription = jest.fn();

export let rejectProtectionRequest = jest.fn();

export function setRejectProtectionRequest(mockFn: any) {
    rejectProtectionRequest = mockFn;
}

export let acceptProtectionRequest = jest.fn();

export function setAcceptProtectionRequest(mockFn: any) {
    acceptProtectionRequest = mockFn;
}

export function decision(legalOfficerAddress: string, decisions: any[]) {
    return decisions[0];
}

export let fetchRecoveryInfo = jest.fn();

export function setFetchRecoveryInfo(mockFn: any) {
    fetchRecoveryInfo = mockFn;
}

export let acceptLocRequest = jest.fn();

export function setAcceptLocRequest(mockFn: any) {
    acceptLocRequest = mockFn;
}

export let rejectLocRequest = jest.fn();

export function setRejectLocRequest(mockFn: any) {
    rejectLocRequest = mockFn;
}

export let addLink = jest.fn();
