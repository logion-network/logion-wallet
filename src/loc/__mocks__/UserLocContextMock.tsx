import { UUID } from "@logion/node-api/dist/UUID";
import { LocData } from "@logion/client/dist/Loc";

import { LocRequest } from "../../common/types/ModelTypes";
import { LocItem } from "../types";

let locState: any;

export function setLocState(value: any) {
    locState = value;
}

export let refresh = jest.fn();

let requestSof = jest.fn();

export function setRequestSof(fn: jest.Mock) {
    requestSof = fn;
}

let requestSofOnCollection = jest.fn();

export function setRequestSofOnCollection(fn: jest.Mock) {
    requestSofOnCollection = fn;
}

let locId: UUID;

export function setLocId(id: UUID) {
    locId = id;
}

let loc: LocData | null = null;

export function setLoc(request: LocData | null) {
    loc = request;
}

let locItems: LocItem[] = [];

export function setLocItems(items: LocItem[]) {
    locItems = items;
}

export let deleteMetadata = jest.fn();

export let deleteFile = jest.fn();

export let addFile = jest.fn().mockResolvedValue(undefined);

export let addMetadata = jest.fn().mockResolvedValue(undefined);

export function useUserLocContext() {

    return {
        refresh,
        locState,
        requestSof,
        requestSofOnCollection,
        locId,
        loc,
        deleteMetadata,
        deleteFile,
        locItems,
        addFile,
        addMetadata,
    }
}
