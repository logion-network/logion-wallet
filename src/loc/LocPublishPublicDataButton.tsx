import { EditableRequest } from "@logion/client";
import { addMetadata, UUID } from "@logion/node-api";

import { LocItem } from "./LocItem";
import LocPublishButton from "./LocPublishButton";
import { publishMetadata } from "src/legal-officer/client";
import { useLogionChain } from "src/logion-chain";

export interface Props {
    locId: UUID;
    locItem: LocItem;
}

export default function LocPublishPublicDataButton(props: Props) {
    const { signer, client } = useLogionChain();

    if(!client) {
        return null;
    }

    return (
        <LocPublishButton
            locItem={ props.locItem }
            publishMutator={ async (current, callback) => {
                if(current instanceof EditableRequest
                    && props.locItem.name
                    && props.locItem.value
                    && props.locItem.submitter) {

                    return await publishMetadata({
                        locState: current,
                        item: {
                            name: props.locItem.name,
                            value: props.locItem.value,
                            submitter: props.locItem.submitter,
                        },
                        signer: signer!,
                        callback,
                    });
                } else {
                    return current;
                }
            }}
            feesEstimator={ () => client.public.fees.estimateWithoutStorage({
                origin: client.currentAddress?.address || "",
                submittable: addMetadata({
                    api: client.nodeApi,
                    locId: props.locId,
                    item: {
                        name: props.locItem.name || "",
                        value: props.locItem.value || "",
                        submitter: props.locItem.submitter!,
                    },
                })
            }) }
            itemType="Public Data"
        />
    )
}
