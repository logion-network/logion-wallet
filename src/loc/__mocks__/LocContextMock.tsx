import { LocData, CollectionItem } from "@logion/client";
import { UUID, ValidAccountId } from "@logion/node-api";

import { LocItem } from "../LocItem";

let locState: any = {
    isLogionIdentity: () => false,
    isLogionData: () => true,
};

export function setLocState(value: any) {
    locState = value;
    if(value.data) {
        loc = {
            ...loc,
            ...locState.data(),
        }
    }
}

let locItems: LocItem[] = [];

export function setLocItems(items: LocItem[]) {
    locItems = items;
}

let locId = new UUID("aed4c6e4-979e-48ad-be6e-4bd39fb94762");

export function setLocId(id: UUID) {
    locId = id;
}

let loc: Partial<LocData> | null = {
    id: locId,
    locType: "Identity",
    requesterAccountId: ValidAccountId.polkadot("5Ew3MyB15VprZrjQVkpQFj8okmc9xLDSEdNhqMMS5cXsqxoW"),
    userIdentity: {
        firstName: "John",
        lastName: "Doe",
        email: "john.doe@logion.network",
        phoneNumber: "+1234",
    }
};

export function setLocRequest(request: Partial<LocData> | null) {
    loc = request;
}

export let refresh = jest.fn();

export let requestSof = jest.fn();

export function setRequestSof(fn: jest.Mock) {
    requestSof = fn;
}

export let deleteMetadata = jest.fn();

export let deleteLink = jest.fn();

let collectionItems: CollectionItem[] = [];

export function setCollectionItems(items: CollectionItem[]) {
    collectionItems = items;
}

async function mutateLocState(mutator: (state: any) => Promise<any>): Promise<void> {
    await mutator(locState);
}

export function useLocContext() {
    return {
        linkLoc: {

        },
        locId,
        loc,
        locItems,
        refresh,
        requestSof,
        deleteMetadata,
        deleteLink,
        checkResult: { result: "NONE" },
        locState,
        collectionItems,
        mutateLocState,
    };
}
