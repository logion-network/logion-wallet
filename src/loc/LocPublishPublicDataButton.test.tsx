jest.mock("./LocContext");
jest.mock("src/logion-chain");

import { HashString } from "@logion/client";
import { UUID, Hash } from "@logion/node-api";
import { MetadataItem } from "./LocItem";
import { shallowRender } from "../tests";
import LocPublishPublicDataButton from "./LocPublishPublicDataButton";
import { mockValidPolkadotAccountId } from 'src/__mocks__/LogionMock';

describe("LocPublishPublicDataButton", () => {

    const locId = new UUID("62e3ea0e-eee5-4295-819e-ed01b55472f0");
    const locItem = new MetadataItem(
        {
            type: "Data",
            status: "DRAFT",
            submitter: mockValidPolkadotAccountId("data-submitter"),
            timestamp: null,
            newItem: false,
            template: false,
        },
        {
            name: HashString.fromValue("data-name"),
            value: HashString.fromValue("data-value"),
        }
    );

    it("renders", () => {
        const tree = shallowRender(<LocPublishPublicDataButton locItem={ locItem } locId={ locId } />)
        expect(tree).toMatchSnapshot();
    })
})
