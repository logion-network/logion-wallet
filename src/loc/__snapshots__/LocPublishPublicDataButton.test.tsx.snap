// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocPublishPublicDataButton renders 1`] = `
<LocPublishButton
  feesEstimator={[Function]}
  itemType="Public Data"
  locItem={
    MetadataItem {
      "_data": Object {
        "name": HashString {
          "hash": Hash {
            "_bytes": Uint8Array [
              41,
              226,
              51,
              143,
              138,
              156,
              180,
              226,
              73,
              80,
              118,
              113,
              116,
              19,
              27,
              150,
              120,
              98,
              225,
              28,
              85,
              144,
              91,
              143,
              175,
              45,
              157,
              82,
              182,
              142,
              164,
              44,
            ],
          },
          "value": "data-name",
        },
        "value": HashString {
          "hash": Hash {
            "_bytes": Uint8Array [
              24,
              214,
              156,
              3,
              103,
              18,
              141,
              3,
              68,
              107,
              41,
              95,
              132,
              88,
              192,
              205,
              243,
              149,
              181,
              41,
              151,
              107,
              139,
              181,
              220,
              39,
              32,
              37,
              202,
              142,
              205,
              27,
            ],
          },
          "value": "data-value",
        },
      },
      "commonData": Object {
        "newItem": false,
        "status": "DRAFT",
        "submitter": Object {
          "address": "data-submitter",
          "equals": [Function],
          "toKey": [Function],
          "type": "Polkadot",
        },
        "template": false,
        "timestamp": null,
        "type": "Data",
      },
    }
  }
  publishMutator={[Function]}
/>
`;
